edx-solutions/edx-platform
# BB-742 Delete profile images from file storage

This PR introduces a Celery [task](https://github.com/edx-solutions/edx-platform/pull/1359/files#diff-174511fe8141042981a061dc33bb1cf8R17-R27) that deletes profile images for a user from file storage.

It also introduces a new method in [`user_api.accounts.api`](https://github.com/edx-solutions/edx-platform/pull/1359/files#diff-df4681b6dbe6bbc209ddf0f521adb0a6R541-R567) for deleting a user and any related objects. This is added to serve as a common entrypoint for deleting various records that have indirect relationship to the User. 

# Testing instructions

## Profile images

### Recommended settings

1. `CELERY_ALWAYS_EAGER = True`

**`lms.env.json`**
```json
{
...
"DEFAULT_FILE_STORAGE": "storages.backends.s3boto.S3BotoStorage",
...
"PROFILE_IMAGE_BACKEND": {
        "class": "storages.backends.s3boto.S3BotoStorage",
        "options": {
            "base_url": "/media/profile-images/",
            "location": "/edx/var/edxapp/media/profile-images/"
        }
},
...
}

```

**`lms.auth.json`**
```json
{
...
    "AWS_ACCESS_KEY_ID": "your-id",
    "AWS_S3_CUSTOM_DOMAIN": "your-bucket.s3.amazonaws.com",
    "AWS_SECRET_ACCESS_KEY": "your-key",
    "AWS_STORAGE_BUCKET_NAME": "your-bucket",
    "DEFAULT_FILE_STORAGE": "storages.backends.s3boto.S3BotoStorage",
...
}
```

**`lms/envs/private.py`**
```python
MEDIA_ROOT = "/edx/var/edxapp/uploads"
DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
```

### Test steps

1. Login to apros (`http://apros.mcka.local`)
2. Upload a profile image via the "change photo" button in the dropdown from clicking your username in the top-right corner
3. Verify the image(s) have been saved. If you have configured the S3 storage backend, check your S3 bucket. If you have configured the local file storage backend, check in `/edx/var/edxapp/media/profile-images/`
4. Start a Django shell for lms and start the profile image deletion task
   ```
   >>> from openedx.core.djangoapps.profile_images.tasks import *
   >>> delete_profile_images.delay(['mcka_admin_user'])
   ```
   Alternatively, use the `delete_user` api
   ```
   >>> from openedx.core.djangoapps.user_api.accounts.api import delete_user
   >>> user = User.objects.get(username='mcka_admin_user')
   >>> delete_user(user)
   ```
5. Verify the image(s) have been deleted.

## Staff Graded Assignments (on hold)

**NOTE** I have reverted the changes that deletes staff graded assignment files until more clarity is available

### Recommended settings

The same as profile images above

### Test steps

1. Create a test course with the `edx_sga` module enabled and add a unit with the sga component
2. Start the course as a user
3. Upload a file via the sga component
4. Verify the file has been saved. If you have configured the S3 storage backend, check your S3 bucket. If you have configured the local file storage backend, check in `/edx/var/edxapp/uploads/`
5. Start a Django shell for lms and start the profile image deletion task
   ```
   >>> from openedx.core.djangoapps.user_api.accounts.tasks import *
   >>> delete_staff_graded_assignment_files([31])
   ```
   Alternatively, use the `delete_user` api
   ```
   >>> from openedx.core.djangoapps.user_api.accounts.api import delete_user
   >>> user = User.objects.get(username='mcka_admin_user')
   >>> delete_user(user)
   ```
6. Verify the file has been deleted.

# Author notes and concerns
- Deletion of certificates is put on hold
- Deletion of assignments is put on hold
- The storage backend is able to properly handle missing/non-existent images (they are skipped/ignored)
- The [`delete_user`](https://github.com/edx-solutions/edx-platform/pull/1359/files#diff-df4681b6dbe6bbc209ddf0f521adb0a6R541-R567) api method currently takes only one user at a time. This was done for simplicity. Though this would mean that when deleting multiple users, each user would get its own task for deleting the profile images.
- Should retry parameters be specified for the Celery task?
- ~Should the edx-sga deletion be implemented in the edx-sga repo? I wasn't able to find an existing fork of edx-sga so I made the changes here.~

# Related PRs
- https://github.com/edx-solutions/api-integration/pull/197

# To Do
- [x] ~delete assignments (waiting for confirmation if [edx-sga](https://github.com/mitodl/edx-sga/tree/master/edx_sga) is the right project)~
